File: bench/logparse/logparse.go
// Log line parsing: string building, trimming, splitting, prefix tests and
// integer conversion, with the results counted in a map.
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	lines := []string{}
	i := 0
	for i < 4000 {
		lines = append(lines, "  level=info user=u"+strconv.Itoa(i%97)+" ms="+strconv.Itoa(i%31)+"  ")
		i = i + 1
	}

	users := map[string]int{}
	elapsed := 0
	round := 0
	for round < 280 {
		j := 0
		for j < len(lines) {
			fields := strings.Split(strings.TrimSpace(lines[j]), " ")
			k := 0
			for k < len(fields) {
				field := fields[k]
				if strings.HasPrefix(field, "user=") {
					user := strings.TrimPrefix(field, "user=")
					users[user] = users[user] + 1
				}
				if strings.HasPrefix(field, "ms=") {
					value, err := strconv.Atoi(strings.TrimPrefix(field, "ms="))
					if err != nil {
						panic(err)
					}
					elapsed = elapsed + value
				}
				k = k + 1
			}
			j = j + 1
		}
		round = round + 1
	}
	fmt.Println(len(users), elapsed)
}

File: bench/logparse/logparse.lua
-- Log line parsing: string building, trimming, splitting, prefix tests and
-- integer conversion, with the results counted in a table.
local lines = {}
local i = 0
while i < 4000 do
    lines[#lines + 1] = "  level=info user=u" .. tostring(i % 97) .. " ms=" .. tostring(i % 31) .. "  "
    i = i + 1
end

local function trim(s)
    return string.match(s, "^%s*(.-)%s*$")
end

local function split(s, sep)
    local parts = {}
    local start = 1
    while true do
        local at = string.find(s, sep, start, true)
        if at == nil then
            parts[#parts + 1] = string.sub(s, start)
            return parts
        end
        parts[#parts + 1] = string.sub(s, start, at - 1)
        start = at + #sep
    end
end

local function has_prefix(s, prefix)
    return string.sub(s, 1, #prefix) == prefix
end

local users = {}
local total = 0
local elapsed = 0
local round = 0
while round < 280 do
    local j = 1
    while j <= #lines do
        local fields = split(trim(lines[j]), " ")
        local k = 1
        while k <= #fields do
            local field = fields[k]
            if has_prefix(field, "user=") then
                local user = string.sub(field, #"user=" + 1)
                if users[user] == nil then
                    users[user] = 1
                    total = total + 1
                else
                    users[user] = users[user] + 1
                end
            end
            if has_prefix(field, "ms=") then
                elapsed = elapsed + tonumber(string.sub(field, #"ms=" + 1))
            end
            k = k + 1
        end
        j = j + 1
    end
    round = round + 1
end
print(total .. " " .. elapsed)
